加载自定义四足机器人环境

在本教程中，我们将了解如何使用MuJoCo/Ant-v5框架创建四足步行环境，使用模型文件（以.xml结尾），而无需创建新类。

Steps:

Get your MJCF (or URDF) model file of your robot.
获取机器人的MJCF （或URDF ）模型文件。

Create your own model (see the Guide) or,
创建您自己的模型（请参阅指南）或者，

Find a ready-made model (in this tutorial, we will use a model from the MuJoCo Menagerie collection).
找到一个现成的模型（在本教程中，我们将使用MuJoCo Menagerie系列中的模型）。

Load the model with the xml_file argument.
使用xml_file参数加载模型。

Tweak the environment parameters to get the desired behavior.
调整环境参数以获得所需的行为。

Tweak the environment simulation parameters.
调整环境模拟参数。

Tweak the environment termination parameters.
调整环境终止参数。

Tweak the environment reward parameters.
调整环境奖励参数。

Tweak the environment observation parameters.
调整环境观测参数。

Train an agent to move your robot.
训练代理来移动你的机器人。

In [3]:
!pip install "gymnasium[mujoco]

Defaulting to user installation because normal site-packages is not writeable
  Using cached absl_py-2.1.0-py3-none-any.whl.metadata (2.3 kB)
  Using cached importlib_resources-6.4.5-py3-none-any.whl.metadata (4.0 kB)
   ---------------------------------------- 0.0/4.7 MB ? eta -:--:--
   ---------------------- ----------------- 2.6/4.7 MB 13.7 MB/s eta 0:00:01
   ---------------------------------------- 4.7/4.7 MB 13.0 MB/s eta 0:00:00
Using cached absl_py-2.1.0-py3-none-any.whl (133 kB)
   ---------------------------------------- 0.0/559.4 kB ? eta -:--:--
   ---------------------------------------- 559.4/559.4 kB 9.6 MB/s eta 0:00:00
   ---------------------------------------- 0.0/2.4 MB ? eta -:--:--
   ---------------------------------------- 2.4/2.4 MB 13.8 MB/s eta 0:00:00
Using cached importlib_resources-6.4.5-py3-none-any.whl (36 kB)


In [4]:
'''
使用Ant-v5框架的xml_file参数。
'''

import gymnasium
import numpy as np
env = gymnasium.make('Ant-v5', xml_file='./mujoco_menagerie/unitree_go1/scene.xml')

In [ ]:
"""虽然这足以加载模型，但我们需要调整一些环境参数以获得环境所需的行为，目前我们还将显式设置模拟、终止、奖励和观察参数。"""
env = gymnasium.make(
    'Ant-v5',
    xml_file='./mujoco_menagerie/unitree_go1/scene.xml',
    forward_reward_weight=0,
    ctrl_cost_weight=0,
    contact_cost_weight=0,
    healthy_reward=0,
    main_body=1,
    healthy_z_range=(0, np.inf),
    include_cfrc_ext_in_observation=True,
    exclude_current_positions_from_observation=False,
    reset_noise_scale=0,
    frame_skip=1,
    max_episode_steps=1000,
)

# Step 2 - Tweaking the Environment Parameters
## Step 2.1 - Tweaking the Environment Simulation Parameters

感兴趣的参数是frame_skip 、 reset_noise_scale和max_episode_steps


## Step 2.2 - Tweaking the Environment Termination Parameters
终止对于机器人环境来说很重要，可以避免采样“无用”的时间步长。
感兴趣的参数是terminate_when_unhealthy和healthy_z_range 。

## 步骤2.3 - 调整环境奖励参数
感兴趣的参数是forward_reward_weight 、 ctrl_cost_weight 、 contact_cost_weight 、 healthy_reward和main_body 。

## Step 2.4 - Tweaking the Environment Observation Parameters 调整环境观测参数
感兴趣的论点是 include_cfrc_ext_in_observation 和 exclude_current_positions_from_observation 。



# Step 3 - Train your Agent

最后，我们完成了，我们可以使用 RL 算法来训练代理来行走/运行Go1机器人。注意：如果您使用自己的机器人模型遵循本指南，您可能会在训练过程中发现某些环境参数不符合预期，请随时返回步骤 2 并根据需要更改任何内容。